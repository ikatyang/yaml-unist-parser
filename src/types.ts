export interface Point {
  /** 1-based */
  line: number;
  /** 1-based */
  column: number;
  /** 0-based */
  offset: number;
}

export interface Position {
  start: Point;
  end: Point;
}

export interface Node {
  type: string;
  position: Position;
}

export interface Parent extends Node {
  children: Node[];
}

// -----------------------------------------------------------------------------

export interface Content {
  anchor: Null | Anchor;
  tag: Null | VerbatimTag | ShorthandTag | NonSpecificTag;
  /** comments between the node and its tag/anchor */
  middleComments: Comment[];
}

export interface CommentAttachable {
  /** comments in front of the node */
  leadingComments: Comment[];
  /** comments on the same line of the node */
  trailingComments: Comment[];
}

export interface EndCommentAttachable {
  /** comments after the node with greater column */
  endComments: Comment[];
}

export interface YAMLSyntaxError extends SyntaxError {
  source: string;
  position: Position;
}

export type YamlUnistNode =
  | Comment
  | VerbatimTag
  | ShorthandTag
  | NonSpecificTag
  | Anchor
  | Root
  | Document
  | DocumentHead
  | DocumentBody
  | Directive
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | MappingItem
  | MappingKey
  | MappingValue
  | Sequence
  | SequenceItem
  | FlowMapping
  | FlowMappingItem
  | FlowSequence
  | FlowSequenceItem
  | Null;

export type ContentNode =
  | Alias
  | BlockLiteral
  | BlockFolded
  | Plain
  | QuoteSingle
  | QuoteDouble
  | Mapping
  | Sequence
  | FlowMapping
  | FlowSequence
  | Null;

// -----------------------------------------------------------------------------

export interface Null extends Node {
  type: "null";
  /** always -1 */
  position: Position;
}

export interface Comment extends Node {
  type: "comment";
  value: string;
  /** @internal non-enumerable */
  parent?: Extract<YamlUnistNode, CommentAttachable>;
}

export interface Anchor extends Node {
  type: "anchor";
  value: string;
}

export interface VerbatimTag extends Node {
  type: "verbatimTag";
  value: string;
}

export interface ShorthandTag extends Node {
  type: "shorthandTag";
  handle: string;
  suffix: string;
}

export interface NonSpecificTag extends Node {
  type: "nonSpecificTag";
}

export interface Root extends Parent {
  type: "root";
  children: Document[];
  comments: Comment[];
}

export interface Document extends Parent, CommentAttachable {
  type: "document";
  children: [DocumentHead, DocumentBody];
  /** always 0 */
  leadingComments: Comment[];
  /** only attachable on `...` */
  trailingComments: Comment[];
}

export interface DocumentHead extends Parent {
  type: "documentHead";
  children: Array<Comment | Directive>;
}

export interface DocumentBody extends Parent {
  type: "documentBody";
  children: Array<Comment | ContentNode>;
}

export interface Directive extends Node, CommentAttachable {
  type: "directive";
  name: string;
  parameters: string[];
}

export interface Alias extends Node, Content, CommentAttachable {
  type: "alias";
  value: string;
}

export interface BlockValue extends Node, Content, CommentAttachable {
  chomping: "clip" | "keep" | "strip";
  indent: null | number;
  value: string;
  /** comments between indicator and value */
  trailingComments: Comment[];
}

export interface BlockLiteral extends BlockValue {
  type: "blockLiteral";
}

export interface BlockFolded extends BlockValue {
  type: "blockFolded";
}

export interface Plain extends Node, Content, CommentAttachable {
  type: "plain";
  value: string;
}

export interface QuoteValue extends Node, Content, CommentAttachable {
  value: string;
}

export interface QuoteSingle extends QuoteValue {
  type: "quoteSingle";
}

export interface QuoteDouble extends QuoteValue {
  type: "quoteDouble";
}

export interface Mapping extends Parent, Content, CommentAttachable {
  type: "mapping";
  children: MappingItem[];
}

export interface MappingItemBase extends Parent, CommentAttachable {
  /** key-value pair */
  children: [MappingKey | Null, MappingValue | Null];
}

export interface MappingItem extends MappingItemBase {
  type: "mappingItem";
  children: [MappingKey, MappingValue | Null];
}

export interface MappingKey
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingKey";
  children: [ContentNode];
}

export interface MappingValue
  extends Parent,
    CommentAttachable,
    EndCommentAttachable {
  type: "mappingValue";
  children: [ContentNode];
}

export interface Sequence
  extends Parent,
    Content,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequence";
  children: SequenceItem[];
}

export interface SequenceItemBase extends Parent {
  children: [ContentNode];
}

export interface SequenceItem
  extends SequenceItemBase,
    CommentAttachable,
    EndCommentAttachable {
  type: "sequenceItem";
}

export interface FlowCollection extends Parent, Content, CommentAttachable {
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowMapping extends FlowCollection {
  type: "flowMapping";
  children: FlowMappingItem[];
}

export interface FlowMappingItem extends MappingItemBase {
  type: "flowMappingItem";
}

export interface FlowSequence extends FlowCollection {
  type: "flowSequence";
  children: Array<FlowMappingItem | FlowSequenceItem>;
}

export interface FlowSequenceItem extends SequenceItemBase {
  type: "flowSequenceItem";
}
