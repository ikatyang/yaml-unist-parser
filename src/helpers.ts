import { wrap } from "jest-snapshot-serializer-raw";
import { parse } from "./parse";
import { Comment, Node, Position, Root, YamlUnistNode } from "./types";
import { isYAMLError } from "./utils";

export type Arrayable<T> = T | T[];

export type TestCase = TestCaseSingle | TestCaseMulti;
export type TestCaseSelector = (root: Root) => Arrayable<YamlUnistNode>;

export type TestCaseSingle = [string, TestCaseSelector];
export type TestCaseMulti = [string, TestCaseSelector[]];

export function getFirstContent<T extends YamlUnistNode>(): (root: Root) => T;
export function getFirstContent<T extends YamlUnistNode>(root: Root): T;
export function getFirstContent<T extends YamlUnistNode>(root?: Root) {
  return !root
    ? (_root: Root) => _root.children[0].children[1].children[0] as T
    : (root.children[0].children[1].children[0] as T);
}

export interface SnapshotNodeOptions extends StringifyNodeOptions {
  codeFrameMaxHeight?: number;
  selectNodeToStringify?: (node: YamlUnistNode) => YamlUnistNode;
}

export function testCases(
  cases: TestCase[],
  options: SnapshotNodeOptions = {},
) {
  cases.forEach(testCase => {
    const [text, selector] = testCase;
    const root = parse(text);
    const selectNodes = ([] as TestCaseSelector[]).concat(selector);
    selectNodes.forEach(selectNode => {
      const nodes = ([] as YamlUnistNode[]).concat(selectNode(root));
      nodes.forEach(node => {
        test(`${JSON.stringify(text).slice(0, 60)}`, () => {
          expect(wrap(snapshotNode(text, node, options))).toMatchSnapshot();
        });
      });
    });
  });
}

function getNodeDescription(node: YamlUnistNode) {
  return `${node.type} (${[
    `${node.position.start.line}:${node.position.start.column}`,
    `${node.position.end.line}:${node.position.end.column}`,
  ].join(" ~ ")})`;
}

function snapshotNode(
  text: string,
  node: YamlUnistNode,
  options: SnapshotNodeOptions = {},
) {
  const { selectNodeToStringify = (x: YamlUnistNode) => x } = options;
  const stringifiedNode = selectNodeToStringify(node);
  return (
    `${getNodeDescription(node)}\n` +
    codeFrameColumns(text, node.position, options.codeFrameMaxHeight) +
    "\n" +
    (node !== stringifiedNode
      ? `${getNodeDescription(stringifiedNode)}\n`
      : "") +
    stringifyNode(stringifiedNode, options)
  ).replace(/ +$/gm, "");
}

export interface StringifyNodeOptions {
  maxChildrenLevel?: number;
  maxCommentsLevel?: number;
}

function stringifyNode(
  node: YamlUnistNode | Node,
  options: StringifyNodeOptions = {},
): string {
  const attributes = Object.keys(node)
    .filter(key => {
      switch (key) {
        case "type":
        case "position":
        case "children":
        case "comments":
        case "leadingComments":
        case "middleComments":
        case "trailingComments":
        case "endComments":
        case "anchor":
        case "tag":
          return false;
        default:
          return true;
      }
    })
    .map(key => `${key}=${JSON.stringify(node[key as keyof typeof node])}`)
    .sort()
    .map((attribute, index) => (index === 0 ? " " + attribute : attribute))
    .join(" ");
  const propNodes =
    "tag" in node
      ? [node.tag, node.anchor]
          .filter(propNode => propNode.type !== "null")
          .map(propNode => stringifyNode(propNode))
      : [];
  const comments =
    options.maxCommentsLevel === undefined || options.maxCommentsLevel > 0
      ? ([] as string[])
          .concat("leadingComments" in node ? "leadingComments" : [])
          .concat("middleComments" in node ? "middleComments" : [])
          .concat("trailingComments" in node ? "trailingComments" : [])
          .concat("endComments" in node ? "endComments" : [])
          .map(key =>
            // @ts-ignore
            (node[key] as Comment[]).map(
              comment =>
                `<${key.slice(0, -1)} value=${JSON.stringify(comment.value)}>`,
            ),
          )
          .reduce((a, b) => a.concat(b), [])
      : [];
  return "children" in node || comments.length !== 0 || propNodes.length !== 0
    ? `<${node.type}${attributes}>\n${indent(
        ([] as string[])
          .concat(
            propNodes,
            comments,
            (options.maxChildrenLevel === undefined ||
              options.maxChildrenLevel > 0) &&
            "children" in node
              ? (node.children as YamlUnistNode[]).map(childNode =>
                  stringifyNode(childNode, {
                    ...options,
                    ...(options.maxChildrenLevel !== undefined && {
                      maxChildrenLevel: options.maxChildrenLevel - 1,
                    }),
                    ...(options.maxCommentsLevel !== undefined && {
                      maxCommentsLevel: options.maxCommentsLevel - 1,
                    }),
                  }),
                )
              : [],
          )
          .join("\n"),
      )}\n</${node.type}>`
    : `<${node.type}${attributes} />`;
}

function indent(text: string) {
  return text
    .split("\n")
    .map(line => "  " + line)
    .join("\n");
}

function codeFrameColumns(
  text: string,
  position: Position,
  codeFrameMaxHeight = Infinity,
) {
  const lines = text.split("\n").map(line => `${line}¶`);
  const markerLines = lines.map((line, index) => {
    if (index < position.start.line - 1 || index > position.end.line - 1) {
      return "";
    }
    if (index === position.start.line - 1) {
      return (
        " ".repeat(position.start.column - 1) +
        (index === position.end.line - 1 &&
        position.start.column === position.end.column
          ? "~"
          : "^".repeat(
              (index === position.end.line - 1
                ? position.end.column - 1
                : line.length) -
                (position.start.column - 1),
            ))
      );
    } else if (index === position.end.line - 1) {
      return position.end.column === 1 && line === "¶"
        ? "^"
        : "^".repeat(position.end.column - 1);
    } else {
      return "^".repeat(line.length);
    }
  });

  const start = Math.max(0, position.start.line - 1 - codeFrameMaxHeight);
  const end = Math.min(lines.length, position.end.line + codeFrameMaxHeight);

  const gutterWidth = Math.floor(Math.log10(lines.length)) + 1;

  return lines
    .slice(start, end)
    .reduce(
      (reduced, line, index) => {
        const gutter = leftpad(`${index + 1 + start}`, gutterWidth);
        return reduced.concat(
          `${gutter} | ${line.replace(/ /g, "·")}`,
          markerLines[index + start]
            ? `${" ".repeat(gutterWidth)} | ${markerLines[index + start]}`
            : [],
        );
      },
      [] as string[],
    )
    .join("\n");
}

function leftpad(text: string, width: number) {
  return " ".repeat(Math.max(0, width - text.length)) + text;
}

export function testSyntaxError(text: string, message?: string) {
  try {
    parse(text);
    throw new Error("SyntaxError not found");
  } catch (error) {
    if (!isYAMLError(error)) {
      throw error;
    }
    test(message || error.message, () => {
      expect(
        error.message + "\n" + codeFrameColumns(error.source, error.position),
      ).toMatchSnapshot();
    });
  }
}
