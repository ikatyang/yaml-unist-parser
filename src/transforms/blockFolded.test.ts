import { getFirstContent, testCases } from "../helpers";
import { BlockFolded } from "../types";

testCases([
  [">\n  123\n  456\n\n\n", getFirstContent()],
  [">+\n  123\n  456\n\n\n", getFirstContent()],
  [">-\n  123\n  456\n\n\n", getFirstContent()],
  [">1\n  123\n  456\n\n\n", getFirstContent()],
  [">1+\n  123\n  456\n\n\n", getFirstContent()],
  [">1-\n  123\n  456\n\n\n", getFirstContent()],
  ["> # hello\n", root => root],
  ["> # hello \n  123\n  456\n\n\n", root => root],
  [
    "  !!str &anchor \n  # comment  \n>\n  123\n  456\n\n\n",
    [
      getFirstContent(),
      root => getFirstContent<BlockFolded>(root).middleComments[0],
    ],
  ],
]);
