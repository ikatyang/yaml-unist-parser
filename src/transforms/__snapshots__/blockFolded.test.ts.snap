// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"  !!str &anchor \\n  # comment  \\n>\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (3:1 ~ 8:1)
1 | ··!!str·&anchor·¶
2 | ··#·comment··¶
3 | >¶
  | ^^
4 | ··123¶
  | ^^^^^^
5 | ··456¶
  | ^^^^^^
6 | ¶
  | ^
7 | ¶
  | ^
8 | ¶
  | ^
<blockFolded chomping="clip" indent=null value="123 456\\n">
  <handleTag handle="!!" suffix="str" />
  <anchor value="anchor" />
  <middleComment value=" comment  ">
</blockFolded>
`;

exports[`"  !!str &anchor \\n  # comment  \\n>\\n  123\\n  456\\n\\n\\n" 2`] = `
comment (2:3 ~ 2:14)
1 | ··!!str·&anchor·¶
2 | ··#·comment··¶
  |   ^^^^^^^^^^^
3 | >¶
4 | ··123¶
5 | ··456¶
6 | ¶
7 | ¶
8 | ¶
<comment value=" comment  " />
`;

exports[`">+\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (1:1 ~ 6:1)
1 | >+¶
  | ^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockFolded chomping="keep" indent=null value="123 456\\n\\n\\n" />
`;

exports[`">\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (1:1 ~ 6:1)
1 | >¶
  | ^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockFolded chomping="clip" indent=null value="123 456\\n" />
`;

exports[`">-\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (1:1 ~ 6:1)
1 | >-¶
  | ^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockFolded chomping="strip" indent=null value="123 456" />
`;

exports[`">1+\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (1:1 ~ 6:1)
1 | >1+¶
  | ^^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockFolded chomping="keep" indent=1 value="  123\\n  456\\n\\n\\n" />
`;

exports[`">1\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (1:1 ~ 6:1)
1 | >1¶
  | ^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockFolded chomping="clip" indent=1 value="  123\\n  456\\n" />
`;

exports[`">1-\\n  123\\n  456\\n\\n\\n" 1`] = `
blockFolded (1:1 ~ 6:1)
1 | >1-¶
  | ^^^^
2 | ··123¶
  | ^^^^^^
3 | ··456¶
  | ^^^^^^
4 | ¶
  | ^
5 | ¶
  | ^
6 | ¶
  | ^
<blockFolded chomping="strip" indent=1 value="  123\\n  456" />
`;
